src/lib/middleware/with-api-key.ts

```diff
@@ -13,30 +14,62 @@ export const withApiKey: Middleware<
     db: Database
   }
 > = (next) => async (req, res) => {
-  if (req.db == null) {
-    return res
-      .status(500)
-      .end(
-        "The withApiKey middleware requires req.db. Use it with the withDb middleware.",
-      )
+  if (req.headers.authorization == null) {
+    throw new UnauthorizedException({
+      type: "unauthorized",
+      message: "No Authorization header",
+    })
   }
 
   const token = req.headers.authorization?.split("Bearer ")?.[1]
-  if (token == null) return res.status(401).end("Unauthorized")
+  if (token == null) {
+    throw new UnauthorizedException({
+      type: "unauthorized",
+      message: "No token provided",
+    })
+  }
 
-  // TODO: Validate authorization.
-  // If relevant, add the user or the decoded JWT to the request on req.auth.
+  if (token.startsWith("seam_cst1")) {
+    throw new UnauthorizedException({
+      type: "client_session_token_used_for_api_key",
+      message: "A client session token was used instead of an API key",
+    })
+  }
+
+  if (token.startsWith("seam_at")) {
+    throw new UnauthorizedException({
+      type: "access_token_used_for_api_key",
+      message: "An access token was used instead of an API key",
+    })
+  }
+
+  let decodedJwt
+  try {
+    decodedJwt = jwt.decode(token)
+  } catch {}
+  if (decodedJwt != null) {
+    throw new UnauthorizedException({
+      type: "unauthorized",
+      message: "A JWT was used instead of an API key",
+    })
+  }
```

Comment:
This block is incorrect in seam-connect but I fixed it here.
In seam-connect it'll never throw when JWT is provided because the error gets caught: https://github.com/seamapi/seam-connect/blob/486c564a7138bca4cd93ce51d6370bee4405fcbd/lib/middlewares/api-key-auth.ts#L54-L61.
Created an issue: https://github.com/seamapi/seam-connect/issues/9520